Create CONTRIBUTING.md (Contribution Guidelines of the CaC/content Project) (#6069)
ComplianceAsCode/content: CONTRIBUTING.md

ComplianceAsCode Contribution Guidelines
===

> 
> We aim to create a welcoming environment for all contributors regardless of their age, gender, experience, nationality, or employer. However, to foster smooth collaboration we are asking contributors to adhere to the following rules.
> 

Reviewer: ```suggestion

We're so glad you're thinking about contributing to ComplianceAsCode! If you're unsure about anything, just ask -- or submit the issue or pull request to start the conversation! The worst that could happen is you'll politely be asked to change something.

We want to ensure a welcoming environment for all members of the ComplianceAsCode community. Reading the projects CONTRIBUTING policy (you are here), its [LICENSE](https://github.com/ComplianceAsCode/content/blob/master/LICENSE), and its [README](https://github.com/ComplianceAsCode/content/blob/master/README.md), will help you get calibrated.

This project is a public-private partnership with the United States Government as part of its [open source initiatives](https://code.nsa.gov/). The project has welcomed contributions from across the United States Government, private industry, and countries worldwide. Currently, there are four project maintainers who are accountable to ensure a welcoming community and our code of conduct is followed. Should you ever need, please feel free to reach out directly!

- @shawndwells (Shawn Wells // CrowdStrike // shawn.wells@crowdstrike.com)
- @redhatrises  (Gabe Alford // Red Hat // galford@redhat.com)
- @jeffblank (Jeff Blank // U.S. Department of Defense // jeffrey.d.blank@ugov.gov)
- @carlosmmatos (Carlos Matos // Red Hat // cmatos@redhat.com)

```

Reviewer 2: Do we really have to mention the partnership with US Government? It could disturb contributors from other parts of the world, and this partnership is invisible to contributors, or it has been for the past 2 years (i.e. after the shorthand->yaml redesign of the project).

Next, I believe that your list of maintainers is outdated - it contains a person that hasn't contributed to the project for more than five years and isn't even a project member, and four top contributors over the past 2 years are completely missing from the list.
This raises an interesting question how does one become and stay a maintainer. Should it be a title for life?

Author: I think that the list of contributors or maintainers (honorable for-life title or not) simply does not belong to Guidelines.

Reviewer 3: Mentioning the US Government is a must, it is the private part of this public/private partnership. It is the reason that CentOS won't have OSPP or STIG profiles as well as for this project to exist and to continue to exist as is. In addition, the project's output get reviewed by the US Government. Code contributions are not the same thing as project ownership and maintainership.

Reviewer 2: This thread is not resolved, please respond to our questions about maintainers.

Reviewer 2: This subject has been also discussed in another thread, and here is the answer: https://github.com/ComplianceAsCode/content/pull/6069#discussion_r517687131

Issues and Pull Requests
---

> Pull requests can’t be merged by the author, a review from another person is needed. If anybody raises material objections to the pull request, those objections should be addressed before the PR merge.

Reviewer 3: self-merging should be allowed for typos. This is okay to enhance & allow.

Reviewer: ```suggestion
- Two-party review is needed for merging pull requests. This second set of eyes helps review the code for functionality and style, and helps create a shared understanding of why certain things are merged (or not!).
```

Reviewer 2: I disagree with the idea to allow self-merging. You can always find somebody to do a quick review, and typos typically aren't urgent.

Related to @shawndwells suggestion, your formulation suggests that approval of a second party is stronger than material objections. The question of merging a pull request that is believed to be fundamentally flawed shouldn't be a question of two people who want to merge it against one who doesn't - those in favor of the merge should address such objection to show that they understand it.

Reviewer 3: While typos aren't urgent and you disagree, many projects do it for typos. I reduces T&E which is important and needed.

Author: I'm on board with simple typo fixes. I could be frustrating and embarrassing to ask somebody to fix your typo. I would be on board even with small text modifications, but only *if they don't change* the meaning.

> If a pull request has received feedback from reviewers, and the original author hasn’t touched it for more than 14 days, the PR may be closed (it is still possible to re-open it if the author catches up later).
> 

Reviewer: ```suggestion
- Ideally pull request feedback would be addressed within 14 days. This helps us move the project forward and helps prevent the need for major rebasing of pull requests. Everyone goes on vacation or has reasons this won't always be possible. If that's the case, please let us know in the pull request to avoid it being closed for no response. Even if your pull request is closed, worst case you'll need to re-open it upon your return (no code will be lost!).

```

Reviewer 2: The original formulation was indeed too strong. I like your suggestions, maybe consider adding a sentence that if a PR is inactive for no apparent reason for a prolonged period of time, we place a "14 days until we close it" warning as a comment.

Reviewer 3: I would make this 30 days rather than 14 days as it gives more time for vacations and such.

Author: OK, 30 days it is.

> - Ideally pull request feedback would be addressed within 30 days. This helps us move the project forward and helps prevent the need for major rebasing of pull requests. Everyone goes on vacation or has reasons this won't always be possible. If that's the case, please let us know in the pull request to avoid it being closed for no response. Even if your pull request is closed, worst case you'll need to re-open it upon your return (no code will be lost!). If a pull request is inactive for no apparent reason for a prolonged period of time, a "30 days until we close it" warning would be placed as a comment to emphasis its stalled state.

Reviewer 2: ... comment to ~emphasis~ emphasize its stalled state.

Pull requests may be closed immediately only if the change would render the build system, a profile, or a tool broken, and, at the same time, there is no way to improve the PR. Justification and an extensive explanation (or a link to a previous case of such a change) should always be provided as a part of the feedback.

Pull requests should pass continuous integration tests. If the CI run ends with an error, the merge can only be approved by an independent reviewer who has the knowledge of the CI. A waiver justification message is required to be a part of the PR.

> Both issues and pull requests should contain a description that clarifies the matter to a degree that a regular contributor can understand it. Issues and PRs are not a personal TODO list, they are problems and solutions descriptions for the whole community to acknowledge and act upon.

Reviewer 3: ```suggestion
Issues and PRs are not a personal TODO list, they are problems and solutions descriptions for the whole community to acknowledge and act upon.
```

This sentence needs to be removed. Issues can very much be a contributors TODO list for a PR.

Reviewer 4: I think it needs rewording, instead. I think the point isn't that the issues can't look list a list of TODO items, that's fine. The problem is different. I think we actually don't want to have issues that only their author can understand. And we don't want issues that don't contain enough information to be actionable. Instead, we want issues that contributors can easily take and resolve.

Author: Yes, the key word is "personal". Maybe this one would be better:
> Both issues and pull requests should contain a description that clarifies the matter to a degree that a regular contributor can understand it. Please use Issues and PRs to describe problems (and possible solutions) for the whole community to acknowledge and act upon.

What to do with things like #5525 is another problem. On one hand they are part of a GH "project" and, as far as I can see, it was an attempt to build some kind of contribution pipeline. On the other hands it looks abandoned, and also its implementation sucks a bit because the project litters Issues list to the level of being non-browsable by humans.

Reviewer 4: yes, these are inhuman

> It is always better to create a set of small self-contained pull requests rather than a single massive PR, if possible. And if a pull request consists of a sequence of well-defined steps, it is advisable to keep the *one step = one commit* approach.

Reviewer: The spirit of these points is valid however the tone is confrontational. Minor rewording would make these points more approachable.

Author: I'm in with rewording, but let's work on it entry by entry. This diff highlight is pretty useless by itself, I'll close it as resolved to stay focused.

Avoid merge commits in pull requests - use *git rebase* to get rid of them.

Contributing Content
---

> Changes related to one distribution or profile shouldn’t change behavior for other distributions or profiles.

Reviewer 4: ```suggestion
Changes related to one product or profile shouldn’t change behavior for other products or profiles.
```

Author: Okay

> Prevent duplication of code. Use Jinja macros and [check (remediation) templates](/docs/manual/developer_guide.adoc#732-list-of-available-templates). Follow the indentation style. For Python code, follow PEP8. [Developer Guide](/docs/manual/developer_guide.adoc) is your friend.

Reviewer 4: ```suggestion
Prevent duplication of code. Use Jinja macros and [check (remediation) templates](/docs/manual/developer_guide.adoc#732-list-of-available-templates). Follow the indentation style. For Python code, follow [PEP8](https://www.python.org/dev/peps/pep-0008/). [Developer Guide](/docs/manual/developer_guide.adoc) is your friend.
```

Author: Sure, why not?

> Prevent duplication of code. Use Jinja macroses and [check (remediation) templates](/docs/manual/developer_guide.adoc#732-list-of-available-templates). Follow the indentation style. For Python code, follow PEP8. [Developer Guide](/docs/manual/developer_guide.adoc) is your friend.

Reviewer 5: ```suggestion
Prevent duplication of code. Use Jinja macros and [check (remediation) templates](/docs/manual/developer_guide.adoc#732-list-of-available-templates). Follow the indentation style. For Python code, follow PEP8. [Developer Guide](/docs/manual/developer_guide.adoc) is your friend.
```

Author: Yeah, macros already plural

> 

Reviewer 4: What about the SSGTS test scenarios? Should we enforce/advice/recommend writing them? What is the recommendation when a contribution breaks a SSGTS scenario? Unfortunately, this isn't covered by the CI statement because they don't run in CI.

Author: Could you please try to formulate the entry of of this? I'm not sure I understand the problem enough to do it myself.

Reviewer 3: The SSGTS test scenarios should be required to run as part of the CI process. Requiring SSGTS is also a barrier to the project. In fact, the fact that the SSGTS doesn't run as part of the CI process is a huge barrier.

Reviewer 4: Yes, it is a huge problem. At this moment it can be a suggestion but without CI we can't enforce it.

Reviewer 6: > Requiring SSGTS is also a barrier to the project.

Yes, test scenarios can be a barrier, but it helps ensure behavior of the contributed code.

>  In fact, the fact that the SSGTS doesn't run as part of the CI process is a huge barrier.

I agree that not running test scenarios as part of CI is a problem.
But even if we had CI running test scenarios, the contributor  should still add them when touching complex rules.

> At this moment it can be a suggestion but without CI we can't enforce it.

Agree, we can (highly) recommend contributing Test Scenarios for the time being.

> 

Reviewer 3: This is not a contribution requirement. It is a suggestion. This makes the project unapproachable by non-technical contributors.

Reviewer 6: I understand and support trying to make life of non-technical contributors easier.

But if a contribution is adding OVAL or remediations, independent of background, tests should be there to ensure expected behavior.

Reviewer 5: I agree with @yuumasato  I hope I am not wrong when I believe that if someone is able to write a Bash remediation, then they are able to write simple Bash test scenarios.

> Enjoy!

Reviewer 3: The overall tone of this is not outsider friendly. I suggest reworking this to match the tone of https://gist.github.com/briandk/3d2e8b3ec8daf5a27a62

Author: Are you referring to the highlighted part only, or the whole document?

Anyhow, this is exactly where helpful hand of a native English-speaker would be welcome.

Author: > This is not a contribution requirement. It is a suggestion. This makes the project unapproachable by non-technical contributors.

Let's have the discussion in a single thread.

Reviewer 3: > Are you referring to the highlighted part only, or the whole document?

I am referring to the whole document, and actually, will have to be reconciled with what already exists in the docs.

ComplianceAsCode Contribution Guidelines
===

We're so glad you're thinking about contributing to ComplianceAsCode! If you're unsure about anything, just ask -- or submit the issue or pull request to start the conversation! The worst that could happen is you'll politely be asked to change something.

We want to ensure a welcoming environment for all members of the ComplianceAsCode community. Reading the projects CONTRIBUTING policy (you are here), its [LICENSE](https://github.com/ComplianceAsCode/content/blob/master/LICENSE), and its [README](https://github.com/ComplianceAsCode/content/blob/master/README.md), will help you get calibrated.

> This project is a public-private partnership with the United States Government as part of its [open source initiatives](https://code.nsa.gov/). The project has welcomed contributions from across the United States Government, private industry, and countries worldwide. Currently, there are four project maintainers who are accountable to ensure a welcoming community and our code of conduct is followed. Should you ever need, please feel free to reach out directly!

Author: With stressing "public-private partnership with the United States Government" in the contribution guidelines you bring implications about some administrative or supervisory role of it in the project. Which means that there are certain constraints for contributions to the project. It would be great to briefly describe these constraints and (or) give a more direct link (referring to https://code.nsa.gov/ is way too vague).

- @shawndwells (Shawn Wells // CrowdStrike // shawn.wells@crowdstrike.com)
- @redhatrises  (Gabe Alford // Red Hat // galford@redhat.com)
- @jeffblank (Jeff Blank // U.S. Department of Defense // jeffrey.d.blank@ugov.gov)

> - @carlosmmatos (Carlos Matos // Red Hat // cmatos@redhat.com)

Author: This list of four names might be seen as an exhaustive list of persons, who could decide the fate of contributions to the project or provide guidance (again, it is not a generic README file, this guideline is all about contributions). Is it the impression you want to communicate?

Reviewer 4: The project has much more maintainers and everyone should be responsible for following this document therefore there is no need to pick 4 of them and list them explicitly.

Reviewer 3: While true to some degree, the project maintainers here are ultimately responsible to the US government as trusted stewards of the private relationship. They are the ones held accountable and are able to meet directly to address concerns and issues over the project including what gets merged.

Reviewer 7: So... as a non-US citizen, I can't become a "maintainer"? Is this what is being conveyed here? The list of maintainers and the paragraph above this just made this whole document very confusing.
What does being a "maintainer" mean for this project then? Might want to start defining that as well.

Reviewer 3: Good point. That's an incorrect assumption which is hopefully clarified now.

1. It is always better to fix an issue in external upstream open source projects before contributing, fixing, or breaking content here. For example if the `gdm` rpm reports changed permissions when the `rpm_verify_permissions` rule runs, the fix needs to be provided in the `gdm` rpm not by changing or disabling the content. Another example of this is adding a configuration option like to `pam_faillock` and `pam_pwquality` to make it easy for centralized accounts to ignore local settings which may be governed by the centralized account management tool. Once the patches are merged in the upstream project, you should then open a new pull request in this project with the changes. Following this approach not only ensures that it is easier for others to securely configure their systems, but it makes other upstream projects better and easier to use.
1. Ask before you disable or remove content because you feel that they are not security relevant. They might actaully be security relevant in ways that you might not realize. This way we all learn something.
1. Products or content that have been previously retired should stay retired unless there is a heavy commitment to support that content.

> 1. While this is an open source project, contributing content from projects that have been taken over by a hacker is a no-go. Not only can you or the project not guarantee that that project is truly cleaned up, but doing so diminishes the trust and provenance of this project.

Reviewer 2: I feel that this paragraph is trying to say something, but I have no idea what. When one contributes rule descriptions, checks and remediations, those should be backed up by tests, and they get an independent review, so the source shouldn't matter, and it may not even be known.
What are other relevant scenarios here that this statement aims to address?

> 1. Both DISA and NSA have said CentOS does not meet the expectations of U.S. Government security requirements. Therefore, there are no government profiles for CentOS, and they are not supported by this project whether through developement or contributions. This clearly angers some, but good cyber practices and compliance are practiced and expected by this project.

Reviewer 2: It's nice to have this information publicly available, although I feel that a README or FAQ would be even a better fit. What would we say to people that would like to contribute a CentOS profile that is not related to US Govt?
